spec/java/security/SecureRandom.main.lsl

```diff
@@ -0,0 +1,885 @@
+        val curServiceType: String = action CALL_METHOD(curService, "getType", []);
+        val curServiceAlgorithm: String = action CALL_METHOD(curService, "getAlgorithm", []);
+
+        if (action OBJECT_EQUALS(curServiceType, "SecureRandom") && action OBJECT_EQUALS(curServiceAlgorithm, _algorithm))
+        {
+            resultAlgorithm = curServiceAlgorithm;
+            action LOOP_BREAK();
+        }
+    }
+
+
+    @throws(["java.security.NoSuchAlgorithmException"])
+    @static fun *.getInstance (_algorithm: String, provider: Provider): SecureRandom
+    {
+        if (provider == null)
+            _throwIAE();
+
+        val providersListLength: int = 1;
+        val providersList: array<Provider> = action ARRAY_NEW("java.security.Provider", providersListLength);
+        providersList[0] = provider;
+        var resultProvider: Provider = null;
+        var resultAlgorithm: String = null;
+
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, providersListLength, +1,
+            findAlgorithmByAllProviders_loop(i, providersList, _algorithm, resultProvider, resultAlgorithm)
+        );
+
+        if (resultAlgorithm == null)
+            _throwNSAE();
+
+        val isDefaultProvider: boolean = _isDefaultProvider(resultProvider);
+
+        result = new SecureRandomAutomaton(state = Initialized,
+            provider = resultProvider,
+            algorithm = resultAlgorithm,
+            defaultProvider = isDefaultProvider,
+        );
+    }
+
+
+    @throws(["java.security.NoSuchAlgorithmException", "java.security.NoSuchProviderException"])
+    @static fun *.getInstance (_algorithm: String, providerName: String): SecureRandom
+    {
+        if (providerName == null || action CALL_METHOD(providerName, "length", []) == 0)
+            _throwIAE();
+
+        val providersList: array<Provider> = action CALL_METHOD(null as Security, "getProviders", []);
+        val providersListLength: int = action ARRAY_SIZE(providersList);
+        var resultProvider: Provider = null;
+        var resultAlgorithm: String = null;
+
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, providersListLength, +1,
+            findAlgorithmByAllProvidersByName_loop(i, providersList, _algorithm, resultProvider, resultAlgorithm, providerName)
+        );
+
+        if (resultProvider == null)
+            _throwNSPE();
+
+        if (resultAlgorithm == null)
+            _throwNSAE();
+
+        val isDefaultProvider: boolean = _isDefaultProvider(resultProvider);
+
+        result = new SecureRandomAutomaton(state = Initialized,
+            provider = resultProvider,
+            algorithm = resultAlgorithm,
+            defaultProvider = isDefaultProvider,
+        );
+    }
+
+
+    @Phantom proc findAlgorithmByAllProvidersByName_loop (i: int, providersList: array<Provider>, algorithm: String, resultProvider: Provider, resultAlgorithm: String, providerName: String): void
```

Comment:
@dog-m 
This loop was deleted, because were used:

```
action SYMBOLIC("java.security.Provider");
action SYMBOLIC("java.lang.String");
```

temporary decision;